cli/cluster/cluster.go

```diff
@@ -0,0 +1,171 @@
+package cluster
+
+import (
+	"fmt"
+	"strings"
+)
+
+const (
+	fmtErrNoPath = "a configuration file or etcd endpoint must be set"
+)
+
+// findInstance finds an instance with the name in the config and returns
+// it full configuration merged from scopes: global, group, replicaset,
+// instance or nil.
+func findInstance(config *Config, name string) *Config {
+	var global, group, replicaset, instance *Config
+	var gname, rname string
+
+	global = NewConfig()
+	config.ForEach(nil, func(path []string, value any) {
+		if len(path) >= 1 && path[0] == "groups" {
+			// Cut "groups".
+			path := path[1:]
+			if len(path) >= 1 {
+				if gname != path[0] && instance == nil {
+					gname = path[0]
+					group = NewConfig()
+				}
+				if len(path) >= 2 && path[1] == "replicasets" {
+					// Cut group name + "replicasets".
+					path := path[2:]
+					if len(path) >= 1 {
+						if rname != path[0] && instance == nil {
+							rname = path[0]
+							replicaset = NewConfig()
+						}
+						if len(path) >= 2 && path[1] == "instances" {
+							// Cut replicaset name + "instances".
+							path := path[2:]
+							if len(path) >= 1 && path[0] == name {
+								if instance == nil {
+									instance = NewConfig()
+								}
+								instance.Set(path[1:], value)
+							}
+						} else if rname == path[0] {
+							replicaset.Set(path[1:], value)
+						}
+					}
+				} else if gname == path[0] {
+					group.Set(path[1:], value)
+				}
+			}
+		} else {
+			global.Set(path, value)
+		}
+	})
+
+	if instance == nil {
+		return nil
+	}
+
+	instance.Merge(replicaset)
+	instance.Merge(group)
+	instance.Merge(global)
+
+	return instance
+}
+
+// HasInstance returns true if an instance with the name exists in the config.
+func HasInstance(config *Config, name string) bool {
+	return findInstance(config, name) != nil
+}
+
+// Instantiate returns a fetched instance config from the cluster config. If
+// the cluster config has the instance then it returns a merged config of the
+// instance from scopes: global, group, replicaset, instance. If the cluster
+// config has not the instance then it returns a global scope of the cluster
+// config.
+func Instantiate(config *Config, name string) *Config {
+	iconfig := findInstance(config, name)
+	if iconfig != nil {
+		return iconfig
+	}
+
+	iconfig = NewConfig()
+	config.ForEach(nil, func(path []string, value any) {
+		if len(path) < 1 || path[0] != "groups" {
+			iconfig.Set(path, value)
+		}
+	})
+
+	return iconfig
+}
+
+// GetClusterConfig returns a cluster configuration according to options. The
+// function returns a cluster config as is, without merging of settings
+// from scopes: global, group, replicaset, instance.
+func GetClusterConfig(opts PathOpts) (*Config, error) {
```

Comment:
I think we must omit PathOpts, since etcd options are a part of cluster configuration: empty, or got from a file.

Reply:
Replaced by `path string`.